agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +92,174 @@ class PythonRefAgentConnection : AgentConnection {
-
-    fun popMessageOfType(type: String): String? {
-        synchronized(receivedMessages) {
-            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
-            if (result != null)
-                receivedMessages.remove(result)
-
-            return result
+    private fun sendJson(obj: Any) = webSocket.sendJson(obj)
+
+    /**
+     * Retrieves a new invite from the agent
+     */
+    override fun generateInvite(): Single<String> {
+        return Single.create { observer ->
+            try {
+                /**
+                 * to generate the invite, send "generate_invite" message to the agent, and wait for "invite_generated"
+                 * which must contain an invite
+                 */
+                webSocket.sendJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE))
+                webSocket.receiveMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).subscribe({ msg ->
+                    observer.onSuccess(msg.invite)
+                }, { e: Throwable -> throw(e) })
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
     }
 
-    fun waitForMessageOfType(type: String): String {
-        var messageOfType: String? = null
-        while (messageOfType == null) {
-            sleep(500)
-            messageOfType = popMessageOfType(type)
+    /**
+     * Waits for incoming connection from whatever party which possesses the invite
+     */
+    override fun waitForInvitedParty(invite: String): Single<IndyPartyConnection> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * On an incoming connection request, the agent must send the "request_received"
+                     */
+                    webSocket.receiveMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).subscribe({
+                        /**
+                         * on the "request_received" message, reply with "send_response" with remote DID to any incoming connection request.
+                         * at this stage it doesn't matter which party sent the connection request
+                         */
+                        sendJson(RequestSendResponseMessage(it.did))
+                        /**
+                         * Check if the STATE already has the public key ("connection_key") corresponding to the invite
+                         */
+                        sendJson(StateRequest())
+                        webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe { stateResponse ->
+                            val pairwise = stateResponse["content"]["pairwise_connections"].find { node ->
+                                node["metadata"]["connection_key"].asText() == getPubkeyFromInvite(invite)
```

Comment:
Could we get pub key once and not parse it in loop?

Reply:
Please double check, it gets decoded only once, there's no loop.

Reply:
there is 
`find`
which is for loop